src/core/mesh/qgsmeshlayer.cpp

```diff
@@ -1503,8 +1503,13 @@ QString QgsMeshLayer::decodedSource( const QString &source, const QString &provi
   QString src( source );
   if ( provider == QLatin1String( "mdal" ) )
   {
-    src = context.pathResolver().readPath( src );
+    QVariantMap uriParts = QgsProviderRegistry::instance()->decodeUri( QStringLiteral( "mdal" ), source );
```

Comment:
I'm thinking this should be sufficiently generic to remove the "provider == "mdal" " condition above (ie. If the uriParts contains a "path" value then we always resolve that). That would help avoid the meshlayer version of decodedSource becoming a horrible nightmare of cludges like QgsVectorLayer::decodedSource has become.

Reply:
Not sure it will help to understand the logic for other provider, for now, only memory mesh provider that will be treated as the MDAL provider.
I don't like sending ALL the logic of treatment of the uri  to the MDAL provider.
But putting the variable `provider` or `mProvider` in `QgsProviderRegistry::instance()->decodeUri()` and then testing if "path" is in `uriParts` should be well generic, like in the last commit?